### ISSUE TITLE
multi_agent::PolicyLearner has unresolved TODO placeholders

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#4]
## Summary

\`src/multi_agent/learner.rs\` has several TODOs that look like the multi-agent path was wired structurally but never fully completed. These will silently produce wrong training signal if someone tries to use it as-is.

## Specifics

\`\`\`rust
// src/multi_agent/learner.rs:72-76
let buffer = RolloutBuffer::new(
    config.buffer_size,
    1, // num_envs
    4, // obs_dim - TODO: make configurable
);
\`\`\`

\`obs_dim\` is hardcoded to \`4\`. This will be wrong for every env except cartpole.

\`\`\`rust
// src/multi_agent/learner.rs:103-110
// 3. Compute advantages
// TODO: Fix API - compute_advantages needs last_values
// self.buffer.compute_advantages(&last_values, self.config.gamma,
// self.config.gae_lambda);
let last_values = vec![0.0]; // Placeholder
self.buffer.compute_advantages(
    &last_values,
    self.config.gamma as f32,
    self.config.gae_lambda as f32,
);
\`\`\`

\`last_values\` is hardcoded to \`vec![0.0]\` — GAE will be computed against a zero bootstrap regardless of where the episode actually ended. For non-terminal truncations this systematically biases advantages.

## Recommended fix

1. Take \`obs_dim\` from \`Environment::observation_space()\` (or pass through \`LearnerConfig\`).
2. Have the simulator send the final \`V(s_{T+1})\` along with the terminal experience so the learner can bootstrap correctly.
3. Until both are fixed, mark \`multi_agent::PolicyLearner\` as experimental in the README so users don't trust the numbers.

## Repro

\`cargo doc --no-deps --document-private-items\` won't surface this. Visible by reading the file.

**Comment by rjwalters** [comment]
## Curator notes (verified against `master`)

Confirmed the issue against `src/multi_agent/learner.rs` on `master` (commit `7ba44e3`). The original report is accurate, and the file actually contains **four** TODO/placeholder defects, not two. All four must be addressed to make `multi_agent::PolicyLearner` produce a correct training signal.

### Inventory of defects

| # | Location | Defect |
|---|----------|--------|
| 1 | `learner.rs:73` | `obs_dim` is hardcoded to `4` in `RolloutBuffer::new(...)`. Correct only for CartPole. |
| 2 | `learner.rs:108` | `last_values = vec![0.0]` placeholder feeds GAE — non-terminal truncations get a zero bootstrap, systematically biasing advantages downward. |
| 3 | `learner.rs:148-149` | `// TODO: Fix - buffer doesn't have capacity() method` — `target_len` is set to `config.buffer_size`, which is fine, but the comment indicates the author didn't notice that `RolloutBuffer::len()` already gives transition count. Stale TODO, but worth resolving while in the file. |
| 4 | `learner.rs:154-160` | **Most severe.** The `Experience` received from the simulator is bound as `_exp` and discarded. `self.buffer.add(...)` is commented out, so the buffer is never populated. `self.buffer.len()` is always `0`, which means: (a) `min_batch_size` guard at line 100 always fails, and (b) the training loop at `train()` never actually runs. The "training loop" is a no-op. Tests don't catch this because `test_learner_creation` only checks construction. |

So in addition to the bias issues the issue describes, the current state is worse: **the learner does not train at all**. It silently spins.

### Root cause / context

- `Experience` (`src/multi_agent/messages.rs:13`) carries `Tensor` observations, an `i64` action, `f32` reward, `value`, `log_prob`, plus `terminated`/`truncated` flags.
- `RolloutBuffer::add` (`src/buffer/rollout/storage.rs:107`) expects `step: usize, env_id: usize, observation: &[f32], action, reward, value, log_prob, terminated, truncated`. Adapting requires:
  1. Tracking a monotonically-increasing `step` index per env (mod `num_steps`).
  2. Converting the `Tensor` observation to a `&[f32]` (e.g. `Vec::<f32>::try_from(&exp.observation)?`).
  3. Choosing `env_id = 0` for the single-env learner, or threading per-env indices if the design later supports multiple envs per learner.
- For the bootstrap value (defect #2), the cleanest path is to compute `V(s_{T+1})` on the learner side: `MlpPolicy::forward(obs)` already returns `(logits, value)` (`src/policy/mlp.rs:174`). The simulator forwards `next_observation` in every `Experience`, so the learner can run the current policy on the *last* `next_observation` of the rollout under `no_grad` and use that as `last_values[env_id]`. Alternatively, extend `Experience` (or add a `FinalValue` message) so the simulator sends the precomputed bootstrap.
- For defect #1, `obs_dim` should come from `LearnerConfig` (set by the caller using `Environment::observation_space().shape()`). The `Population::new` constructor already takes `obs_dim` (`src/multi_agent/population.rs:29`), so the call site already has it — just thread it through.

### Recommended implementation approach

1. **Add `obs_dim: usize` to `LearnerConfig`.** Update `LearnerConfig::default()` to a sentinel or require explicit construction (default `4` is fine for backward compat with existing tests; document the contract). Use it at `learner.rs:73`.
2. **Wire `RolloutBuffer::add` in `collect_experiences`.** Maintain a `step_idx: usize` field on `PolicyLearner` that increments on each accepted experience, wraps at `config.buffer_size`, and resets in `train()` after `buffer.reset()`. Convert `exp.observation` to `Vec<f32>` via `Vec::<f32>::try_from(&exp.observation)?`.
3. **Bootstrap GAE correctly.** Keep the last received `Experience` per rollout (its `next_observation` and `terminated` flag). At step 3 of `train()`:
   - If the rollout ended on a terminal step (`terminated == true`), `last_values = vec![0.0]`.
   - Otherwise (truncation or just buffer-full), compute `last_value = policy.forward(&next_obs).1` under `no_grad`, then `last_values = vec![last_value.double_value(&[]) as f32]`.
4. **Remove stale TODO comments** at `learner.rs:148-149` and `:155-159`.
5. **Add a regression test** that constructs a learner, pushes a handful of `Experience` values into the experience channel, and verifies after one training cycle that `buffer.len() > 0` and at least one `TrainingStats` has been produced (use a small `buffer_size`/`min_batch_size` in the test).
6. **README:** Until items 1-3 land, add a one-line note under the multi-agent section that `multi_agent::PolicyLearner` is experimental.

### Acceptance criteria

- [ ] `LearnerConfig` exposes `obs_dim` (or equivalent), and `PolicyLearner::new` uses it when constructing `RolloutBuffer`. No hardcoded `4`.
- [ ] `collect_experiences` calls `self.buffer.add(...)` with correctly-translated `Experience` fields; experiences are no longer dropped.
- [ ] GAE bootstrap value reflects `V(s_{T+1})` for non-terminal rollout ends; uses `0.0` only when `terminated == true`.
- [ ] All four TODO/placeholder comments in `src/multi_agent/learner.rs` are resolved (removed or replaced with finalized code).
- [ ] Unit test demonstrates that experiences flow into the buffer and a training step executes when the channel is fed.
- [ ] Existing tests (`test_learner_config_default`, `test_learner_creation`) still pass.
- [ ] README documents multi-agent learner status accurately for the new state.

### Test plan

- `cargo test --lib multi_agent::learner` — existing tests plus the new buffer-flow test.
- `cargo test --lib buffer::rollout` — guard against regressions in the GAE path.
- Manual: run a short multi-agent training session (e.g. modify an existing example or write a 50-step smoke loop) and confirm: (a) `tracing` logs report non-zero `policy_loss` / `value_loss`, (b) advantages are non-trivial (not all near-zero), (c) no panics on tensor shape mismatches when `obs_dim != 4` (CartPole is `4`, but try Snake's larger obs).
- `cargo clippy --all-targets -- -D warnings` — should pass; in particular, no `unused_variables` warning on `_exp` once it's actually used.

### Code references

- `src/multi_agent/learner.rs:70-74` (obs_dim TODO)
- `src/multi_agent/learner.rs:104-113` (last_values placeholder)
- `src/multi_agent/learner.rs:145-171` (collect_experiences — the silent no-op)
- `src/multi_agent/messages.rs:13-72` (`Experience` shape and `is_done()` helper)
- `src/buffer/rollout/storage.rs:107-130` (`RolloutBuffer::add` signature)
- `src/buffer/rollout.rs:42-44` (`compute_advantages` API)
- `src/policy/mlp.rs:174` (`MlpPolicy::forward` returns `(logits, value)` — needed for bootstrap)
- `src/multi_agent/population.rs:29` (population already takes `obs_dim` — proves call site has it)

### Suggested labels

`tier:goal-supporting` — this is correctness-critical infrastructure for the multi-agent training mode rather than a milestone deliverable on its own.
